File: Anul II/Semestrul II/Medii de proiectare si programare/AgentiiZboruri/Persistence/src/main/java/ro/persistence/AngajatDbRepo.java
package ro.persistence;

import ro.model.Angajat;
import ro.model.Entity;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Map;
import java.util.Optional;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.*;

public class AngajatDbRepo implements Repository {

    private final DbUtils dbUtils;

    public AngajatDbRepo(DbUtils dbUtils) {
        this.dbUtils = dbUtils;
    }

    @Override
    public Optional findOne(Object o) {
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("SELECT a.id, a.username, a.password, p.nume, p.prenume FROM Angajat a JOIN Persoana p ON a.id = p.id WHERE a.id=?")) {
            preStmt.setInt(1, (Integer) o);
            try (ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    Integer id = result.getInt("id");
                    String username = result.getString("username");
                    String password = result.getString("password");
                    String nume = result.getString("nume");
                    String prenume = result.getString("prenume");
                    Angajat angajat = new Angajat(id, nume, prenume, username, password);
                    return Optional.of(angajat);
                }
            }
        } catch (SQLException ex) {
            System.out.println("Error DB " + ex);
        }
        return Optional.empty();
    }

    public Optional findOneU(String username) {
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("SELECT a.idP, a.username, a.password, p.nume, p.prenume FROM Angajat a JOIN Persoana p ON a.idP = p.id WHERE a.username=?")) {
            preStmt.setString(1, username);
            try (ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    Integer id = result.getInt("idP");
                    String password = result.getString("password");
                    String nume = result.getString("nume");
                    String prenume = result.getString("prenume");
                    Angajat angajat = new Angajat(id, nume, prenume, username, password);
                    return Optional.of(angajat);
                }
            }
        } catch (SQLException ex) {
            System.out.println("Error DB " + ex);
        }
        return Optional.empty();
    }

    @Override
    public Iterable findAll() {
        Connection con = dbUtils.getConnection();
        List<Angajat> angajati = new ArrayList<>();
        try (PreparedStatement preStmt = con.prepareStatement("SELECT a.id, a.username, a.password, p.nume, p.prenume FROM Angajat a JOIN Persoana p ON a.id = p.id")) {
            try (ResultSet result = preStmt.executeQuery()) {
                while (result.next()) {
                    Integer id = result.getInt("id");
                    String username = result.getString("username");
                    String password = result.getString("password");
                    String nume = result.getString("nume");
                    String prenume = result.getString("prenume");
                    Angajat angajat = new Angajat(id, nume, prenume, username, password);
                    angajati.add(angajat);
                }
            }
        } catch (SQLException ex) {
            System.out.println("Error DB " + ex);
        }
        return angajati;
    }

    @Override
    public Optional save(Entity entity) {
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("INSERT INTO Angajat (id, username, password) VALUES (?, ?, ?)")) {
            Angajat angajat = (Angajat) entity;
            preStmt.setInt(1, angajat.getId());
            preStmt.setString(2, angajat.getUsername());
            String encryptedPassword = encryptPassword(angajat.getPassword());
            preStmt.setString(3, encryptedPassword);
            preStmt.executeUpdate();
        } catch (SQLException ex) {
            System.out.println("Error DB " + ex);
            return Optional.of(entity);
        }
        return Optional.empty();
    }

    public boolean login(String username, String password) {
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("SELECT password FROM Angajat WHERE username = ?")) {
            preStmt.setString(1, username);
            try (ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    String storedPassword = result.getString("password");
                    String encryptedPassword = encryptPassword(password);
                    return storedPassword.equals(encryptedPassword);
                }
            }
        } catch (SQLException ex) {
            System.out.println("Error DB " + ex);
        }
        return false;
    }

    public String encryptPassword(String password) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] hash = md.digest(password.getBytes());
            StringBuilder sb = new StringBuilder();
            for (byte b : hash) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Optional delete(Object o) {
        return Optional.empty();
    }

    @Override
    public Optional update(Entity entity) {
        return Optional.empty();
    }

    @Override
    public Iterable changeEntities(Map entities) {
        return null;
    }
}

File: Anul II/Semestrul II/Medii de proiectare si programare/AgentiiZboruriC/repository/AngajatDbRepo.cs
using System.Data.SQLite;
using System.Security.Cryptography;
using System.Text;
using AgentiiZboruriC.domain;
using log4net.Repository.Hierarchy;

namespace AgentiiZboruriC.repository;

using System;
using System.Collections;
using System.Collections.Generic;
using System.Data.SqlClient;
using log4net;

public class AngajatDbRepo : Repository<int, Angajat>
    {
        private readonly DbUtils dbUtils;

        public AngajatDbRepo(DbUtils dbUtils)
        {
            this.dbUtils = dbUtils;
        }

        public Angajat? FindOne(object o)
        {
            using (SQLiteConnection con = dbUtils.GetConnection())
            using (SQLiteCommand cmd = new SQLiteCommand("SELECT a.id, a.username, a.password, p.nume, p.prenume FROM Angajat a JOIN Persoana p ON a.id = p.id WHERE a.id=@id", con))
            {
                cmd.Parameters.AddWithValue("@id", (int)o);
                using (SQLiteDataReader result = cmd.ExecuteReader())
                {
                    if (result.Read())
                    {
                        int id = result.GetInt32(0);
                        string username = result.GetString(1);
                        string password = result.GetString(2);
                        string nume = result.GetString(3);
                        string prenume = result.GetString(4);
                        return new Angajat(id, nume, prenume, username, password);
                    }
                }
            }
            return null;
        }

        public Angajat? FindOneU(string username)
        {
            using (SQLiteConnection con = dbUtils.GetConnection())
            using (SQLiteCommand cmd = new SQLiteCommand("SELECT a.idP, a.username, a.password, p.nume, p.prenume FROM Angajat a JOIN Persoana p ON a.idP = p.id WHERE a.username=@username", con))
            {
                cmd.Parameters.AddWithValue("@username", username);
                using (SQLiteDataReader result = cmd.ExecuteReader())
                {
                    if (result.Read())
                    {
                        int id = result.GetInt32(0);
                        string password = result.GetString(1);
                        string nume = result.GetString(2);
                        string prenume = result.GetString(3);
                        return new Angajat(id, nume, prenume, username, password);
                    }
                }
            }
            return null;
        }

        public Zbor FindOne(int id)
        {
            throw new NotImplementedException();
        }

        public IEnumerable<Angajat> FindAll()
        {
            List<Angajat> angajati = new List<Angajat>();
            using (SQLiteConnection con = dbUtils.GetConnection())
            using (SQLiteCommand cmd = new SQLiteCommand("SELECT a.id, a.username, a.password, p.nume, p.prenume FROM Angajat a JOIN Persoana p ON a.id = p.id", con))
            {
                using (SQLiteDataReader result = cmd.ExecuteReader())
                {
                    while (result.Read())
                    {
                        int id = result.GetInt32(0);
                        string username = result.GetString(1);
                        string password = result.GetString(2);
                        string nume = result.GetString(3);
                        string prenume = result.GetString(4);
                        angajati.Add(new Angajat(id, nume, prenume, username, password));
                    }
                }
            }
            return angajati;
        }

        public Optional<Angajat> Save(Angajat entity)
        {
            using (SQLiteConnection con = dbUtils.GetConnection())
            using (SQLiteCommand cmd = new SQLiteCommand("INSERT INTO Angajat (id, username, password) VALUES (@id, @username, @password)", con))
            {
                cmd.Parameters.AddWithValue("@id", entity.Id);
                cmd.Parameters.AddWithValue("@username", entity.Username);
                cmd.Parameters.AddWithValue("@password", EncryptPassword(entity.Password));
                cmd.ExecuteNonQuery();
            }
            return null;
        }

        public Optional<Angajat> Delete(int id)
        {
            throw new NotImplementedException();
        }

        public bool Login(string username, string password)
        {
            using (SQLiteConnection con = dbUtils.GetConnection())
            using (SQLiteCommand cmd = new SQLiteCommand("SELECT password FROM Angajat WHERE username = @username", con))
            {
                cmd.Parameters.AddWithValue("@username", username);
                using (SQLiteDataReader result = cmd.ExecuteReader())
                {
                    if (result.Read())
                    {
                        string storedPassword = result.GetString(0);
                        return storedPassword.Equals(EncryptPassword(password));
                    }
                }
            }
            return false;
        }

        public string EncryptPassword(string password)
        {
            using (SHA256 sha256Hash = SHA256.Create())
            {
                byte[] bytes = sha256Hash.ComputeHash(Encoding.UTF8.GetBytes(password));
                StringBuilder builder = new StringBuilder();
                for (int i = 0; i < bytes.Length; i++)
                {
                    builder.Append(bytes[i].ToString("x2"));
                }
                return builder.ToString();
            }
        }

        public Angajat? Delete(object o)
        {
            return null;
        }

        public Optional<Angajat> Update(Angajat entity)
        {
            return null;
        }

        public IEnumerable<Angajat> ChangeEntities(Dictionary<int, Angajat> entities)
        {
            return null;
        }
    }